File: Day 13/solution.cpp
#include <filesystem>
#include <fstream>
#include <iostream>
#include <vector>
#include <cassert>
#include <functional>

typedef std::pair<int, int> coordinate;

class Solver
{
	std::filesystem::path filepath;

	std::vector<std::string> get_one_graph(std::ifstream &file)
	{
		assert(!file.eof() && "File reached EOF");

		std::vector<std::string> result;
		for (std::string line; std::getline(file, line) && line != "";)
		{
			result.push_back(line);
		}
		return result;
	}

	std::vector<std::string> transpose(const std::vector<std::string> &graph)
	{
		std::vector<std::string> result(graph.front().size());
		for (const std::string &row : graph)
		{
			for (int i = 0; i < row.size(); ++i)
			{
				result[i] += row[i];
			}
		}
		return result;
	}

	int generic_solve(std::function<bool(std::vector<std::string>, int)> is_valid_reflection)
	{
		int row_total = 0, col_total = 0;
		std::ifstream file(this->filepath);
		int count = 0;
		while (!file.eof())
		{
			std::vector<std::string> graph = this->get_one_graph(file);
			bool found_reflection = false;

			// Check vertical reflection
			for (int j = 0; j < graph.front().size() - 1; ++j)
			{
				if (is_valid_reflection(graph, j))
				{
					col_total += j + 1;
					found_reflection = true;
					++count;
					break;
				}
			}
			if (found_reflection)
				continue;

			// Check horizontal reflection
			graph = this->transpose(graph);
			for (int i = 0; i < graph.front().size() - 1; ++i)
			{
				if (is_valid_reflection(graph, i))
				{
					row_total += i + 1;
					++count;
					break;
				}
			}
		}
		file.close();
		return col_total + 100 * row_total;
	}

	int part_1()
	{
		auto is_valid_reflection = [](const std::vector<std::string> &reflection_sequences, int reflection_point) -> bool
		{
			for (const std::string &sequence : reflection_sequences)
			{
				for (int i = 0; i < std::min(reflection_point + 1, (int)sequence.size() - reflection_point - 1); ++i)
				{
					if (sequence[reflection_point - i] != sequence[reflection_point + i + 1])
					{
						return false;
					}
				}
			}
			return true;
		};
		return this->generic_solve(is_valid_reflection);
	}

	int part_2()
	{
		auto is_valid_reflection = [](const std::vector<std::string> &reflection_sequences, int reflection_point) -> bool
		{
			int count = 0;
			for (const std::string &sequence : reflection_sequences)
			{
				for (int i = 0; i < std::min(reflection_point + 1, (int)sequence.size() - reflection_point - 1); ++i)
				{
					if (sequence[reflection_point - i] == sequence[reflection_point + i + 1])
						continue;
					if (++count > 1)
						return false;
				}
			}
			return count == 1;
		};
		return this->generic_solve(is_valid_reflection);
	}

public:
	Solver(std::filesystem::path filepath = "input.txt") : filepath(filepath) {}

	void solve()
	{
		std::cout << "Day 13" << std::endl;
		std::cout << "Solving: " << this->filepath << std::endl;
		std::cout << "Part 1:" << std::endl;
		std::cout << this->part_1() << std::endl;
		std::cout << "---" << std::endl;
		std::cout << "Part 2:" << std::endl;
		std::cout << this->part_2() << std::endl;
	}
};

int main(int argc, char *argv[])
{
	std::unique_ptr<Solver> solver;
	if (argc == 1)
	{
		solver = std::make_unique<Solver>();
	}
	else if (argc == 2)
	{
		solver = std::make_unique<Solver>(argv[1]);
	}
	else if (argc == 3 && (argv[1] == "-p" || argv[1] == "--path"))
	{
		solver = std::make_unique<Solver>(argv[2]);
	}
	else
	{
		return 1;
	}

	solver->solve();
}

File: Day 13/solution.py
"""
Day 13 solution.
"""
import sys
from typing import Callable, Generator, Iterable


class Solver:
    """Day 13 solver."""

    def __init__(self, filepath: str = "input.txt"):
        self.filepath = filepath

    def graph_generator(self) -> Generator[list[str], None, None]:
        """
        Parse the input file and generate one graph at a time.
        """
        with open(
            self.filepath, "r", encoding=sys.getdefaultencoding()
        ) as file:
            line = file.readline()
            while line != "":
                graph = []
                while line not in ["\n", ""]:
                    graph.append(line.strip())
                    line = file.readline()

                yield graph
                line = file.readline()

    def generic_solve(
        self, is_valid_reflection: Callable[[list[Iterable[str]], int], bool]
    ) -> int:
        """Generic solve."""
        row_total = col_total = 0
        for graph in self.graph_generator():
            found_reflection = False

            # Check vertical reflection
            for j in range(len(graph[0]) - 1):
                if is_valid_reflection(graph, j):
                    col_total += j + 1
                    found_reflection = True
                    break
            if found_reflection:
                continue

            # Check horizontal reflection
            for i in range(len(graph) - 1):
                if is_valid_reflection(zip(*graph), i):
                    row_total += i + 1
                    break

        return col_total + 100 * row_total

    def part_1(self) -> int:
        """Part 1 solver."""

        def is_valid_reflection(
            reflection_sequences: list[Iterable[str]], reflection_point: int
        ) -> bool:
            return all(
                all(
                    sequence[reflection_point - i]
                    == sequence[reflection_point + i + 1]
                    for i in range(
                        min(  # Get the smaller of the left or right side
                            reflection_point + 1,
                            len(sequence) - reflection_point - 1,
                        )
                    )
                )
                for sequence in reflection_sequences
            )

        return self.generic_solve(is_valid_reflection)

    def part_2(self) -> int:
        """Part 2 solver."""

        def is_valid_reflection(
            sequences: list[Iterable[str]], reflection_point: int
        ) -> bool:
            return (
                sum(
                    sum(
                        sequence[reflection_point - i]
                        != sequence[reflection_point + i + 1]
                        for i in range(
                            min(
                                reflection_point + 1,
                                len(sequence) - reflection_point - 1,
                            )
                        )
                    )
                    for sequence in sequences
                )
                == 1
            )

        return self.generic_solve(is_valid_reflection)

    def solve(self) -> None:
        """Runs part 1 and part 2."""
        print("Day 13")
        print(f"Solving: {self.filepath}")
        print("Part 1:")
        print(self.part_1())
        print("---")
        print("Part 2:")
        print(self.part_2())


if __name__ == "__main__":
    match sys.argv:
        case [_]:
            solver = Solver()
        case [_, "--path", path] | [_, "-p", path] | [_, path]:
            solver = Solver(path)
        case _:
            print("Usage: python solution.py <path>")
            sys.exit(0)
    solver.solve()
